server/src/main/java/org/elasticsearch/snapshots/RestoreService.java

```diff
@@ -510,6 +513,30 @@ private void startRestore(
         );
     }
 
+    private void validateDatastreamTemplatesExistAndWarnIfMissing(Map<String, DataStream> dataStreamsToRestore, SnapshotInfo snapshotInfo) {
+        var templatePatterns = clusterService.state()
+            .metadata()
+            .templatesV2()
+            .values()
+            .stream()
+            .filter(cit -> cit.getDataStreamTemplate() != null)
+            .flatMap(cit -> cit.indexPatterns().stream())
+            .collect(Collectors.toSet());
```

Comment:
I think you are right... There might be a few failed rollover attempts while the restore is in progress (if the DS gets restored before the tamplates) but I believe it would rollover successfully in the end so it's a valid scenario and should not warn.

I'll get that logic in and add some tests for these senario.